Widgets/CodeWidgetScintilla.cpp

```diff
@@ -1,27 +1,60 @@
 #include "CodeWidget.h"
+#include "Components/ArtManager.h"
 
+#include <Qsci/qsciapis.h>
 #include <Qsci/qscilexercpp.h>
 #include <Qsci/qsciprinter.h>
 #include <Qsci/qsciscintilla.h>
 
 #include <QFontMetrics>
 #include <QLayout>
 #include <QPrintDialog>
+#include <QShortcut>
 
 namespace {
-QsciLexerCPP* cppLexer = nullptr;
-}
 
-CodeWidget::CodeWidget(QWidget* parent) : QWidget(parent), font(QFont("Courier", 10)) {
-  QFontMetrics fontMetrics(font);
+QsciLexerCPP* cppLexer = nullptr;
+QsciAPIs* sciApis = nullptr;
 
+void prepare_scintilla_apis() {
   if (cppLexer == nullptr) {
     cppLexer = new QsciLexerCPP();
-    cppLexer->setFont(font);
+    cppLexer->setFont(QFont("Courier", 10));
+  }
+
+  if (sciApis == nullptr) {
+    sciApis = new QsciAPIs(cppLexer);
   }
+}
+
+}  // anonymous namespace
+
+void CodeWidget::prepareKeywordStore() { prepare_scintilla_apis(); }
+
+void CodeWidget::addKeyword(const QString& keyword) { sciApis->add(keyword); }
+
+void CodeWidget::finalizeKeywords() { sciApis->prepare(); }
+
+CodeWidget::CodeWidget(QWidget* parent) : QWidget(parent), font(QFont("Courier", 10)) {
+  prepare_scintilla_apis();
+
+  QFontMetrics fontMetrics(font);
 
   QsciScintilla* codeEdit = new QsciScintilla(this);
   this->textWidget = codeEdit;
+  codeEdit->registerImage(0, QPixmap(":/actions/right-arrow-red.png"));

```

Comment:
What's going on with the hard-coded numbers? What do they mean? Why isn't this just happening in some kind of loop? Does nothing cache these images? Why are object and sprite assumed to be pre-loaded, but those other two images are not? Don't you have some kind of icon manager for this?

Reply:
1) The numbers thing is QSci's way of tagging keywords, it's the QSci analogy to grpc's tags.
2) Whatever you want them to mean.
3) uwutm8
4) I don't know if anything cache's them, my code doesn't, but QsciScintilla might.
5) I just hacked two resources on there to demonstrate it real fast, I don't actually have it hooked up to show resource keywords just yet. We actually also need a custom derived GML/EDL lexer in order to highlight these keywords, which is pretty easy (just override the keyword method and return them etc).
6) We have ArtManager which I added caching of pixmaps to in the other room editor pr.

But yeah, I'll take some steps to clean this shit up, I think we could use an enum too.

Reply:
Alright: 4c2f5a33d5a03ce4234b8e9a7a1a341b95e3c07c

I've typed the keywords now with an enum which should make that a lot more sensible. There's some work left to be done here. One thing to do is to derive a custom QSciLexer from QSciLexerCPP for GML/EDL so these keywords are also syntax highlighted and things like `with` are considered keywords. We also need to include the resource keywords which is going to be tricky because it doesn't seem possible to update keywords once they are added and prepared. You also apparently can't have more than one `QSciApi` associated with a lexer.